WIP: Configure: rethink target resolution

levitte opened the pull request:
Until now, targets are resolved entirely at every level of
inheritance, and they keys in each target were further process in hash
order.  For quite a lot of values, this works fine.  However, the
interaction bwteen the |thread_scheme| setting and the threads()
function is suboptimal (i.e. mostly not working at all).

There are two possible situations that need resolving:

case 1:

    'target' => {
        cflags		=> threads('-pthread'),
        thread_scheme	=> 'pthreads',
    }

This case is resolved by checking and processing specific settings
(thread_scheme) very early with special hooks that affect other table
entries ($disabled{threads} in this case).

case 2:

    'template' => {
        cflags		=> threads('-pthread'),
        thread_scheme	=> '(unknown)',		# default for disabled
    },
    'target' => {
        inherit_from	=> [ 'template' ],
        thread_scheme	=> 'pthreads',
    }

With the current resolution method, this situation is very tricky, not
to say impossible, because when the |thread_scheme| setting in 'target'
is seen, |threads('-pthread')| has already been processed.

This case is resolved by making inheritance much more lazy, and
basically make all processing of CODErefs happen in a combined fashion
is late as possible, i.e. when 'target' itself is processed.  When
processing inheritance, the inherited values simply "bubble up" and
are combined with the values of the current target.  This happens
recursively.

levitte committed:
[to be squashed] restore process_value

levitte commented:
Note: I'm unsure if this should be backported to 1.1.0, in part or entirely.  Please advice.

dot-asm commented on Configure:
```diff
@@ -1108,25 +1108,18 @@ if (!$disabled{dso} && $target{dso_scheme} ne "")
+        # system-dependent compiler options that are necessary.  We
+        # can't truly check that the given options are correct, but
+        # we expect the user to know what [s]He is doing.
+        if ($no_user_cflags && $no_user_defines) {
```
As already mentioned elsewhere (as a side note), these two are always 0 here, so this condition is never met.

levitte replied:
I missed that...  or have forgotten.  I'll try to remember for another commit.

levitte committed:
Configure: rethink early config resolution checks

Early checks of certain config settings is tightly coupled with
certain configuration options that may or may not be set automatically
depending on the values of corresponding config settings.

Therefore, we configure exactly which config settings should affect
what configuration options, and how.  This simplifies the early config
resolution checks to be one routine that jugles the same hash tables.
If the configuration options have been enabled or disabled manually
(directly or indirectly), the early checks leave them unchanged.

levitte committed:
Configure: fix check of user cflags and user defines

levitte commented:
Alright, I've added some more commits, one to generalise the config resolution early checks, and one to fix the issue with `$no_user_cflags` and `$no_user_defines`

dot-asm commented on Configure:
```diff
@@ -299,7 +299,26 @@ $config{openssldir}="";
+my %auto = (
+    threads => "unavailable",
+    shared => "no-shared-target",
+    pid => "no-shared-target",
```
I think I meant "pic", not "pid"...

levitte replied:
Yes.  Fixed

dot-asm replied:
I don't see it. That it's fixed...

levitte replied:
Oops.  I thought I...?  Doing it now.

dot-asm commented on Configure:
```diff
@@ -650,7 +669,7 @@ while (@argvcopy)
                         }
 		# No longer an automatic choice
-		$auto_threads = 0 if ($1 eq "threads");
+		$auto{$1} = 0;
```
I don't quite follow this, $1 can be anything here... Is it intentional that you set say $auto{zlib}?

levitte replied:
Yes.  For the use now, `$auto{zlib} = 0` says that the zlib option has been configured manually, and well, I don't see the harm.

dot-asm replied:
Trouble with this of course is that it's challenging to tell cases like zlib apart. I mean normally you'd just `if (!$auto{zlib})`, but as it stands now you'd have to `if (defined($auto{zlib}) && !$auto{zlib})`. With this in mind I'd say that it would be appropriate to have some condition, e.g. `$auto{$1} = 0 if (defined($auto{$1}))`. This way at least it would be less confusing as it would indicate that %auto can have only specific keys.

levitte replied:
Yeah, ok, you're right, quite unecessary to cause confusion.  Fixing now.

dot-asm commented on Configure:
```diff
@@ -1360,9 +1359,7 @@ unless ($disabled{"crypto-mdebug-backtrace"})
 
 if ($user_cflags ne "") { $config{cflags}="$config{cflags}$user_cflags"; $config{cxxflags}="$config{cxxflags}$user_cflags";}
-else                    { $no_user_cflags=1;  }
 if (@user_defines) { $config{defines}=[ @{$config{defines}}, @user_defines ]; }
```
What's the point of these two if's now, without else's? I mean if $iser_cflags is empty, appending it would yield same string. I'd also argue in favour of `$config{cflags} .= $user_cflags`...

levitte replied:
Good point.  Fixing.

dot-asm commented:
On *side* note, there are things in `@disable_cascades` that can't be disabled, e.g. md5, sha, rsa. [And correspondingly there are non-disablable things in apps/openssl.c:list_disabled().]

dot-asm commented:
On *side* note I find example suggested in description misleading.
```
'template' => {
    cflags		=> threads('-pthread'),
    thread_scheme	=> '(unknown)',		# default for disabled
},
```
What's rationale exactly? That -pthread is a common compiler flag, which becomes handy on platforms that support pthreads? That one doesn't want to make an assumption whether or not platform served by common compiler supports pthreads? I'd say that in such case it would be more appropriate to arrange
```
gcc-pthreads => {
    template => 1,
    cflags => threads('-pthread'),
    thread_scheme => 'pthreads'
}
```
Because if one doesn't want to make assumption that platform served by common compiler supports threads, then one shouldn't make an assumption that supported thread_scheme is pthreads. We even have such example among supported platforms ;-) gcc serves both linux and mingw, but they support different thread_schemes.

levitte committed:
[to be squashed] be more selective when zeroing an element in %auto

levitte commented:
Regarding that latest note, I agree it's misleading...  I miswrote that commit message, good catch!

levitte commented:
I've changed the case 2 example to this:

```
    'base' => {         # base defaults
        thread_scheme   => '(unknown)',
    },
    'compiler-template' => {
        cflags		=> threads('-pthread'),
    },
    'target' => {
        inherit_from	=> [ 'base', 'compiler-template' ],
        thread_scheme	=> 'pthreads',
    }
```

That should be more satisfactory, and corresponds a lot more closely to what we're facing (well, not quite yet, but I'm coming to that in another PR ;-) )

dot-asm commented:
> I've changed the case 2 example to this:

I'd still argue that it's misleading. Point was that it should rather look as following:
```
    'base' => {         # base defaults
        thread_scheme   => '(unknown)',
    },
    'compiler-template' => {
    },
    'compiler-pthreads-template' => {
        cflags		=> threads('-pthread'),
        thread_scheme	=> 'pthreads',
    },
    'target' => {
        inherit_from	=> [ 'base', 'compiler-template',  'compiler-pthreads-template' ],
    }
```

levitte commented:
I disagree for the commit message.  I see no reason to over complicate that example that's supposed to show the "race condition" between the setting of certain setting (`thread_scheme`) and the conditional setting of another (through `threads('-pthread')`).  I could just as well have written the example like this:

```
    'base' => {         # base defaults
        foo_scheme   => '(unknown)',  # 'foo' is considered disabled with this value
    },
    'compiler-template' => {
        cflags		=> enabled('foo', '-foo=2'),  # enabled() generates '-foo=2' lazilly if 'foo' is enabled
    },
    'target' => {
        inherit_from	=> [ 'base', 'compiler-template' ],
        foo_scheme	=> 'bar',
    }
```

levitte commented:
That being said, I'm taking note of your idea with a separate compiler-threads template.

dot-asm commented:
> I could just as well have written the example like this:

Well, such example would make perfect sense, because it depends on presence of an attribute rather than attribute's  specific value. I mean I would have no problems with compiler template looking as following:
```
template => {
   cflags => is_pthreads("-pthreads")
}
target => {
  inherit_from [ 'template' ],
  pthreads
}
```
Well, attribute value would work as well, as in
```
template => {
   cflags => is_threads_scheme_pthreads("-pthreads")
}
target => {
  inherit_from [ 'template' ],
  threads_scheme = "pthreads"
}
```
Once again, just consider gcc serving both linux and mingw. Common gcc template would have to accommodate different thread_schemes... However!!! It's as important to not overcomplicate things. So that above examples are not actually suggestions. gcc-pthreads-template would solve the problem, and there is no reason to make it more complicated than that.

dot-asm commented:
> It's as important to not overcomplicate things. So that above examples are not actually suggestions.

Which kind of calls for question if referred counter-example was an overcomplication...

levitte commented:
> Well, such example would make perfect sense, because it depends on presence of an attribute rather than attribute's specific value.

That's the case with the threads stuff, you know ;-)

levitte commented:
Anyhow, can we get back to *this* PR? I'm sure we will have more discussions of this sort in PRs to come ;-)

dot-asm commented on Configure:
```diff
@@ -1088,25 +1103,18 @@ if (!$disabled{dso} && $target{dso_scheme} ne "")
+        # we expect the user to know what [s]He is doing.
+        if ($user_cflags ne "" && !@user_defines) {
+            die "You asked for multi-threading support, but didn't\n"
+                ,"provide any system-specific compiler options\n";
```
So that attempt to `./Configure gcc enable-threads` should terminate with this `die` message, right? It doesn't. It should be `$user_cflags eq ""`.

There is "[s]He" typo, i.e. capital H.

[On side note I find above placement of comma in die counter-intuitive. Indentation gives clear enough clue.]

dot-asm replied:
On second *side* note/thought I even wonder if it has to be that way? I mean do we actually want to provide this fall-back? I mean expecting that user specifying enable-threads would provide thread flags on config command line? Especially if we take into consideration that nowadays enable-threads kind of implies that you need scheme, and you can't do that on config command line. In other words shouldn't one instruct users to put together own Configuration/99-myconf.conf irregardless if they passed custom flags or not?

dot-asm replied:
> enable-threads kind of implies that you need thread_scheme

By this I mean that we no longer provide option for application to specify multi-threading callbacks, don't we? So that if user wants to implement own thread_scheme [s]he has to provide own crypto/threads_whatever.c replacement. With this in mind *maybe* thread_scheme should be name of this file... Or am I completely off?

levitte replied:
Actually, you're absolutely right.  This is a remainder from pre-1.1.0 times and cannot be supported as is with 1.1.0 and up.  OpenSSL-ish routines like this provided in `crypto/threads_whatever.c` are required.

A curiosity (and side note), btw...  it seems like with have screwed up with unixware.  The thread scheme there is 'uithreads', the C flag is `-Kthreads`, but we don't really have support for it.  As a matter of fact, the way `crypto/threads_whatever.c` are currently written, we assume pthreads, which I can only assume is faulty like hell...  except for `unixware-7-gcc`

dot-asm replied:
> screwed up with unixware

We should really discard it and few others [we never had hands-on experience with]. Suggested strategy [if anybody asks/cares] should be that legacy OS might as well run legacy release.

dot-asm replied:
> This is a remainder from pre-1.1.0 times and cannot be supported as is with 1.1.0 and up.

So you'll remove this or fix it up? I mean if not removed there is wrong condition to fix... [And typo].

levitte replied:
I intend to change this to a simple no-can-do warning.

levitte committed:
[to be squashed] adjust threads check to early config resolution checks

this also greatly simplifies the display of warnings of options that were
forcibly disabled because the feature isn't available

levitte commented:
I threw away the previous somewhat quirky handling of threads settings.  I hope that answers your concerns, @dot-asm

levitte commented:
... and now that I've done this, I'm not entirely sure it's the right choice after all....  consider what's written in INSTALL:

```
 On other systems, to enable support for multi-threading, you will have
 to specify at least two options: "threads", and a system-dependent option.
 (The latter is "-D_REENTRANT" on various systems.)  The default in this
 case, obviously, is not to include support for multi-threading (but
 you can still use "no-threads" to suppress an annoying warning message
 from the Configure script.)
```

(INSTALL also says that we only support two models, pthread and Windows threads, so whatever user flags that are given must add support for one of those models)

Thoughts?

levitte committed:
[to be squashed] still allow users to enable threads of the add compiler flags to support them

dot-asm commented:
> so whatever user flags that are given must add support for one of those models)

Recall that you can't specify thread_scheme on config command line, so it becomes one big contradiction. I mean if there is platform config that specifies thread_scheme, then it's more than appropriate to specify even corresponding flags there, there is no reason to leave it to user. And if platform config doesn't specify thread_scheme, no user flag can possibly solve the problem, because user can't override thread_scheme. On in other words. User should not have an option to specify compile flags on config command line, or should be able to choose even thread_scheme at command line. I really see no reason for latter, because it's easy to put together own Configuration/80-myplatform.conf.

> Thouhgts?

INSTALL describes something that doesn't actually work.

[And as already mentioned, it might be appropriate if thread_scheme was name of a file to compile, as apposite to pair of chosen recognized strings. Idea is to allow developer to provide custom "thread-er" as part of Configuration/80-myplatform.conf. Otherwise [s]he would be forced to mess with Configure, and we don't really want that.]

richsalz commented:
Perhaps we could could do something like we did with the RAND seeding?  So we could have --with-threads={native,none,external} ?

dot-asm commented:
> User should not have an option to specify compile flags on config command line

I naturally meant that user should not have an option to specify *threading* compile flags. Well, we can't really tell threading from non-threading apart, so we can't prevent user from doing so. But it doesn't mean that we should suggest that user can *solve* threading support by adding flags to config command line.

levitte commented:
Oh, I dunno...  consider the case when pthread isn't natively supported, but there is an add-on library that can be built in a specific location, so configuring with `-I/opt/pthread/include -L/opt/pthread/lib` would be a possibility.  Why not trust the user in this case?

dot-asm commented:
> Why not trust the user in this case?

If we *assume* that user-supplied compiler flags passed at command line are meant to engage only *one* threading_scheme, then we can trust user. I mean if `./Configure platform threads -I/opt/something/include -L/opt/something/lib` can only mean "use pthreads for thread_scheme", then sure, we can trust user. *Though* not more than we trust user to understand and accept that pthreads is the *only* option here. Is it reasonable assumption?

Flipping the question, it is rather what's more reasonable: a) creating an illusion that *every* kind of configuration problems can be solved at config command line for *any* platform [including unseen one!]; or b) instructing user to put together own .conf file? Wouldn't we be better off if user-supplied compiler flags passed at config time were used for customization of *recognized/supported* platforms? And new/unseen platforms were handled by Configuration/\*.conf entries?

levitte commented:
> Flipping the question, it is rather what's more reasonable: a) creating an illusion that *every* kind of configuration problem~s~ can be solved at config command line for *any* platform [including unseen one!]; or b) instructing user to put together own .conf file? Wouldn't we be better off if user-supplied compiler flags passed at config time were used for customization of *recognized/supported* platforms? And new/unseen platforms were handled by Configuration/\*.conf entries?

This confuses me...  we have a number of config targets with `thread_scheme => "(unknown)"`...  I'm thinking that if someone has a pthread library for that target, why should they have to create a whole new config target?  Or do you see thread support as a separate platform?

INSTALL also says that we only support two models, pthread and Windows threads.  If they want to add another model, then there's coding to do.

Most of all, though, *this* PR isn't primarly about our thread support.  I do have some thoughts about that, but am reserving that for *another* PR...  and I'm not sure that dramatic changes of this support is ok mid 1.1.x.

levitte commented:
It's true, though, that we're not displaying very well what threading model we expect for specific targets.  The easy answer is that we expect Windows threads on a restricted set of platforms, and pthread everywhere else.  However, changing that is also a subject for *another* PR.

dot-asm commented:
> This confuses me... we have a number of config targets with thread_scheme => "(unknown)"... I'm thinking that if someone has a pthread library for that target, why should they have to create a whole new config target?

And question why shouldn't they is as legitimate. Is it prohibitively expensive or something? Also note that they are free to submit request that would change thread_scheme from "(unknown)" to "pthreads" (or "winthreads"). But think about this. It would be only appropriate to ask if this works for all platform versions. And in case it doesn't, we might insist on two configs, one for contemporary and for old one (provided that keeping legacy config is considered desirable). Would you find such setup overly expensive and hard to manage and maintain? In the essence it all boils down to how to present a platform, as slightly customizable config target, or as something vaguely defined that allows for practically infinite customization of virtually each aspect. You seem to suggest latter, while what I'm saying is that it would mean too much of hardly justified complexity. [And I have deja-vu feeling that we already had this discussion...]

dot-asm commented:
> Most of all, though, this PR isn't primarly about our thread support.

But it's primarily about how its configuration is handled. As per description. So I don't quite follow what is it I say is off topic? Well, I can accept that suggestion to make thread_scheme a compilable file is a bit off topic, but it does underpins my point. And point is that thread_scheme and corresponding flags belong together, either in platform config or template. There is lesser point to break them apart, e.g. to implied "pthreads" and user-supplied compiler options passed at config time.

levitte commented:
> In the essence it all boils down to how to present a platform, as slightly customizable config target, or as something vaguely defined that allows for practically infinite customization of virtually each aspect. You seem to suggest latter, ...

Actually, getting to say `./config threads -I/what/ever/include -L/what/ever/lib` is a slight customization in my mind...  so no, I wouldn't say that my suggestion goes toward infinite customization or virtually each aspect.

The thing is, and I feel that I'm repeating what I've already said, *this is something we currently say we support* (and I would fully expect might slight customization to work already in master), and *I'm unsure it's ok to change this support mid 1.1.x*

levitte commented:
> And point is that thread_scheme and corresponding flags belong together, either in platform config or template. There is lesser point to break them apart, e.g. to implied "pthreads" and user-supplied compiler options passed at config time.

Yes, I got that point.  I'll point out that I changed the message in that particular commit again, so we don't keep getting side tracked.

levitte commented:
> But it's primarily about how its configuration is handled. 

It's primarly about how config resolution is handled, and first and foremost how inheritance is treated.

dot-asm commented:
> Actually, getting to say ./config threads -I/what/ever/include -L/what/ever/lib is a slight customization in my mind...

The only problem with that is that it's suggested that it implies that thread_scheme is pthreads. And then question is if it's universally applicable assumption? If we decide that it is, then no problem. But in such case it would have to be explicitly said so and coded accordingly. I.e. somewhere in Configure there should be line that assigns thread_scheme to "pthreads".

levitte commented:
> The only problem with that is that it's suggested that it implies that thread_scheme is pthreads. And then question is if it's universally applicable assumption? If we decide that it is, then no problem.

We have already decided that it is (unless winthreads), see INSTALL, at least for now (note that I'm discussing this in #4732)

> But in such case it would have to be explicitly said so and coded accordingly. I.e. somewhere in Configure there should be line that assigns thread_scheme to "pthreads".

The way it currently works, `thread_scheme` works like a flag, where `"(unknown)"` means `no-thread` as default, any other string means that we support whatever we decided on the platform, as decided by the code in `crypto/threads_*.c`, i.e. there's currently no correlation whatsoever between the value of `thread_scheme` (except for `"(unknown)"`) and what we actually implement. (note that I'm discussing this in #4732)

dot-asm commented:
Ah! It looks like I fell victim to common-sense expectations. I mean I was kind of assuming that thread_scheme value is meaningful, when it's nothing like that... Well, you've got to admit it does send quite mixed signals... I still think that it's somewhat counter-intuitive and misleading to imply that any user-supplied flag qualifies for pthread support, and that we would be better off if threads capability was required to be part of platform config... Apologize for kind of detour!

levitte commented:
Awright.  ((pfew))

So how are we on *this* PR?

dot-asm commented:
> So how are we on this PR?

I'm looking at it, it obviously doesn't go as fast as you might like... And I still might bring up some issues that might appear as not spot-on related... Basically what I'm trying to grasp now if it's actually necessary to single out specific attributes for treating them that specially... I mean shouldn't there be generic solution? But it's my problem, not yours... I mean don't answer this, just give me time...

levitte commented:
The attributes that are getting special treatment are those that directly affect `%disabled`...  those need to get their final value and `%disabled` need to get its correct values before any function that checks the corresponding `%disabled` entries are performed.  Since this all happens within `resolve_config`, that's where this early resolution of any attribute of this type is needed.

levitte commented:
Hmmm...  I have screwed up this branch, commits seem to be missing.  Ah well, I'm rethinking this effort anyway, so perhaps I should just as well start over...

levitte commented:
Ah, no, not as screwed up as I first thought...  the first commit's message is screwed up, though.

In the mean time, I'm marking this WIP

levitte commented:
I'm replacing this with #4854
